src/backend/distributed/master/master_metadata_utility.c

```diff
@@ -589,25 +621,28 @@ NodeHasActiveShardPlacements(char *nodeName, int32 nodePort)
 	SysScanDesc scanDescriptor = NULL;
 	ScanKeyData scanKey[scanKeyCount];
 
-	Relation pgShardPlacement = heap_open(DistShardPlacementRelationId(),
-										  AccessShareLock);
+	uint32 groupId = GroupForNode(nodeName, nodePort);
 
-	ScanKeyInit(&scanKey[0], Anum_pg_dist_shard_placement_nodename,
-				BTEqualStrategyNumber, F_TEXTEQ, CStringGetTextDatum(nodeName));
-	ScanKeyInit(&scanKey[1], Anum_pg_dist_shard_placement_nodeport,
-				BTEqualStrategyNumber, F_INT8EQ, Int64GetDatum(nodePort));
-	ScanKeyInit(&scanKey[2], Anum_pg_dist_shard_placement_shardstate,
-				BTEqualStrategyNumber, F_INT4EQ, Int32GetDatum(FILE_FINALIZED));
+	Relation pgPlacement = heap_open(DistPlacementRelationId(),
+									 AccessShareLock);
 
-	scanDescriptor = systable_beginscan(pgShardPlacement,
-										DistShardPlacementNodeidIndexId(), indexOK,
+	ScanKeyInit(&scanKey[0], Anum_pg_dist_placement_groupid,
+				BTEqualStrategyNumber, F_INT4EQ, UInt32GetDatum(groupId));
+	if (onlyConsiderActivePlacements)
+	{
+		ScanKeyInit(&scanKey[1], Anum_pg_dist_placement_shardstate,
+					BTEqualStrategyNumber, F_INT4EQ, Int32GetDatum(FILE_FINALIZED));
+	}
+
+	scanDescriptor = systable_beginscan(pgPlacement,
+										DistPlacementGroupidIndexId(), indexOK,
 										NULL, scanKeyCount, scanKey);
 
 	heapTuple = systable_getnext(scanDescriptor);
 	hasFinalizedPlacements = HeapTupleIsValid(heapTuple);
 
 	systable_endscan(scanDescriptor);
-	heap_close(pgShardPlacement, AccessShareLock);
+	heap_close(pgPlacement, AccessShareLock);
```

Comment:
I think the reason for this change to `NoLock` is that callers of `NodeHasShardPlacements` are going to take action based on the result and it would be nice if we didn't `DROP pg_dist_placement` or make any drastic changes while this transaction is still running?

This makes sense, but now I'm confused why `heap_close` includes a parameter to drop locks, it seems you would _never_ want to drop locks until the transaction ends.